### ISSUE TITLE
Random large delays between task runs

### ISSUE THREAD

**Issue by daledude** [intelsdi-x/snap#1575]
I have 4 tasks (dbi, processes, procstat, psutil) that are configured to run every 60 seconds. I'm using the default config given within redhat rpm. snap version 1.2 on kernel 2.6.18-238.9.1.el5. Below is a clip of the snap log where it shows nothing run between 03:43 and 03:48.

```
time="2017-03-28T03:42:57-07:00" level=debug msg="Ping received" _module=plugin-exec io=stderr plugin=influxdb 
time="2017-03-28T03:43:02-07:00" level=debug msg="Ping received" _module=plugin-exec io=stderr plugin=influxdb 
time="2017-03-28T03:43:02-07:00" level=debug msg="Ping received" _module=plugin-exec io=stderr plugin=processes 
time="2017-03-28T03:43:02-07:00" level=debug msg="Ping received" _module=plugin-exec io=stderr plugin=procstat 
time="2017-03-28T03:43:02-07:00" level=debug msg="Ping received" _module=plugin-exec io=stderr plugin=influxdb 
time="2017-03-28T03:43:02-07:00" level=debug msg="Ping received" _module=plugin-exec io=stderr plugin=psutil 
time="2017-03-28T03:43:02-07:00" level=debug msg="Ping received" _module=plugin-exec io=stderr plugin=influxdb 
time="2017-03-28T03:43:07-07:00" level=debug msg="Ping received" _module=plugin-exec io=stderr plugin=processes 
time="2017-03-28T03:43:07-07:00" level=debug msg="Ping received" _module=plugin-exec io=stderr plugin=influxdb 
time="2017-03-28T03:43:07-07:00" level=debug msg="Ping received" _module=plugin-exec io=stderr plugin=procstat 
time="2017-03-28T03:43:07-07:00" level=debug msg="Ping received" _module=plugin-exec io=stderr plugin=psutil 
time="2017-03-28T03:43:07-07:00" level=debug msg="Ping received" _module=plugin-exec io=stderr plugin=influxdb 
time="2017-03-28T03:43:07-07:00" level=debug msg="Ping received" _module=plugin-exec io=stderr plugin=influxdb 
time="2017-03-28T03:48:09-07:00" level=debug msg="Starting workflow" _block=workflow-start _module=scheduler-workflow task-id=61688ecc-772f-43c1-a892-19b3509e9c30 task-name=psutil 
time="2017-03-28T03:48:09-07:00" level=debug msg="starting collector job" _module=scheduler-job block=run job-type=collector metric-count=17 
time="2017-03-28T03:48:09-07:00" level=debug msg="Starting workflow" _block=workflow-start _module=scheduler-workflow task-id=adab00b0-4414-47be-bc8e-3265f5aa117d task-name=processes 
time="2017-03-28T03:48:09-07:00" level=debug msg="starting collector job" _module=scheduler-job block=run job-type=collector metric-count=3 
time="2017-03-28T03:48:09-07:00" level=debug msg="Starting workflow" _block=workflow-start _module=scheduler-workflow task-id=ef29afd8-f2e8-4787-a9b8-69a7dc2be3c5 task-name=procstat 
time="2017-03-28T03:48:09-07:00" level=debug msg="starting collector job" _module=scheduler-job block=run job-type=collector metric-count=1 
time="2017-03-28T03:48:09-07:00" level=debug msg="Starting workflow" _block=workflow-start _module=scheduler-workflow task-id=89ceea6c-78c7-4fe0-bf05-fcc0997b66fd task-name=dbi 
time="2017-03-28T03:48:09-07:00" level=debug msg="starting collector job" _module=scheduler-job block=run job-type=collector metric-count=8 
time="2017-03-28T03:48:09-07:00" level=debug msg="cache miss [/intel/psutil/vm/buffers:9]" _module=routing-cache hits=0 misses=253 namespace="/intel/psutil/vm/buffers:9" 
time="2017-03-28T03:48:09-07:00" level=debug msg="cache miss [/intel/psutil/vm/total:9]" _module=routing-cache hits=0 misses=253 namespace="/intel/psutil/vm/total:9" 
time="2017-03-28T03:48:09-07:00" level=debug msg="cache miss [/intel/psutil/vm/used:9]" _module=routing-cache hits=0 misses=253 namespace="/intel/psutil/vm/used:9" 
time="2017-03-28T03:48:09-07:00" level=debug msg="cache miss [/intel/psutil/cpu/cpu-total/softirq:9]" _module=routing-cache hits=0 misses=253 namespace="/intel/psutil/cpu/cpu-total/softirq:9" 
time="2017-03-28T03:48:09-07:00" level=debug msg="cache miss [/intel/psutil/cpu/cpu-total/user:9]" _module=routing-cache hits=0 misses=253 namespace="/intel/psutil/cpu/cpu-total/user:9" 

```

**Comment by IzabellaRaulin** [comment]
Hello @daledude, I checked how it looks on my setup - I tested task with cpu collector and file publisher. I received "ping" from those two plugins at intervals of 5 s:
```
DEBU[2017-03-28T14:06:32+02:00] Ping received   _module=plugin-exec io=stderr plugin=cpu
DEBU[2017-03-28T14:06:32+02:00] Ping received  _module=plugin-exec io=stderr plugin=file
DEBU[2017-03-28T14:06:37+02:00] Ping received   _module=plugin-exec io=stderr plugin=file
DEBU[2017-03-28T14:06:37+02:00] Ping received  _module=plugin-exec io=stderr plugin=cpu
DEBU[2017-03-28T14:06:42+02:00] Ping received   _module=plugin-exec io=stderr plugin=file
DEBU[2017-03-28T14:06:42+02:00] Ping received  _module=plugin-exec io=stderr plugin=cpu

DEBU[2017-03-28T14:06:43+02:00] Starting workflow   _block=workflow-start _module=scheduler-workflow task-id=8e824c26-19ce-4124-8f57-5a6138181170 task-name=Task-8e824c26-19ce-4124-8f57-5a6138181170
DEBU[2017-03-28T14:06:43+02:00] starting collector job  _module=scheduler-job block=run job-type=collector metric-count=24
```

In this case below starting a workflow started on 14:06:43 (so only 1 sec after the previous log), but I can image that starting workflow might happen **up to 5 sec after the last logged received ping** based on how it works. In the meantime, nothing special happens - the task is just waiting for the right moment to fire determined by an interval. 

@daledude - could you confirm that 
- ping receiving is logged every 5 seconds from each of plugins
- the log "Starting workflow" appears after the bunch of "Ping received" logs **no later than 5s**

To sum up, that is how it works now. @daledude, do you have any suggestion what might be improved in this area? Or is it ok for you?

**Comment by IzabellaRaulin** [comment]
Oh, now I see that between the last "Ping Received" log and "Starting workflow" you got **5 minutes gap**. What is the status of your task after that? Do you have any Task Hits/Misses or Failures (you can check it in output `snaptel task list`? Basically, your task has 60s interval, so it's strange that there is no log in the meantime.

**Comment by daledude** [comment]
@IzabellaRaulin , below is the list output:
```
ID 					 NAME 						 STATE 		 HIT 	 MISS 	 FAIL 	 CREATED 		 LAST FAILURE
2c429883-5ddb-4236-8e8a-50c3aa525856 	 dbi                                       	 Running 	 95 	 0 	 0 	 4:00AM 3-28-2017 	                                                                                    
c8c1e8ef-2d1d-497b-bb44-94cbd18bae2f 	 processes                                 	 Running 	 95 	 0 	 0 	 4:00AM 3-28-2017 	                                                                                    
33343e9e-8b92-4093-8fed-0c4b6203d12e 	 procstat                                  	 Running 	 95 	 0 	 2 	 4:00AM 3-28-2017 	 Worker refused to run overdue job.                                                 
4600371f-5902-491e-925d-77d9d8b6f996 	 psutil                                    	 Running 	 95 	 0 	 0 	 4:00AM 3-28-2017 	                                                                                    
```

Here is a snippet from the /tmp/snap_plugin.log:
```
>>>2017/03/28 03:43:02 Heartbeat timeout reset
>>>2017/03/28 03:43:07 Ping received
>>>2017/03/28 03:48:09 Ping received
>>>2017/03/28 03:48:09 CollectMetrics called
>>>2017/03/28 03:48:12 Ping received
>>>2017/03/28 03:48:12 Heartbeat timeout reset
```

Looking for another delay I came across the below which shows a **Heartbeat timeout**. Is my assumption that a single plugin shouldn't delay any other?

```
time="2017-03-28T06:53:03-07:00" level=debug msg="Ping received" _module=plugin-exec io=stderr plugin=procstat 
time="2017-03-28T06:53:03-07:00" level=debug msg="Ping received" _module=plugin-exec io=stderr plugin=psutil 
time="2017-03-28T06:53:03-07:00" level=debug msg="Ping received" _module=plugin-exec io=stderr plugin=influxdb 
time="2017-03-28T06:53:03-07:00" level=debug msg="Ping received" _module=plugin-exec io=stderr plugin=influxdb 
time="2017-03-28T06:58:07-07:00" level=debug msg="Heartbeat timeout 1 of 3.  (Duration between checks 10s)" _module=plugin-exec io=stderr plugin=psutil 
time="2017-03-28T06:58:08-07:00" level=debug msg="Starting workflow" _block=workflow-start _module=scheduler-workflow task-id=2c429883-5ddb-4236-8e8a-50c3aa525856 task-name=dbi 
time="2017-03-28T06:58:08-07:00" level=debug msg="starting collector job" _module=scheduler-job block=run job-type=collector metric-count=8 
time="2017-03-28T06:58:08-07:00" level=debug msg="Starting workflow" _block=workflow-start _module=scheduler-workflow task-id=c8c1e8ef-2d1d-497b-bb44-94cbd18bae2f task-name=processes 
time="2017-03-28T06:58:08-07:00" level=debug msg="starting collector job" _module=scheduler-job block=run job-type=collector metric-count=3 
time="2017-03-28T06:58:08-07:00" level=debug msg="Starting workflow" _block=workflow-start _module=scheduler-workflow task-id=33343e9e-8b92-4093-8fed-0c4b6203d12e task-name=procstat 
time="2017-03-28T06:58:08-07:00" level=debug msg="starting collector job" _module=scheduler-job block=run job-type=collector metric-count=1 
time="2017-03-28T06:58:08-07:00" level=debug msg="Starting workflow" _block=workflow-start _module=scheduler-workflow task-id=4600371f-5902-491e-925d-77d9d8b6f996 task-name=psutil 
time="2017-03-28T06:58:08-07:00" level=debug msg="starting collector job" _module=scheduler-job block=run job-type=collector metric-count=17 

```

**Comment by mkuculyma** [comment]
@daledude could you provide us your task manifests and global config (with [setfile](https://github.com/intelsdi-x/snap-plugin-collector-dbi#configuration-and-usage) for dbi collector) so that we could try to reproduce the problem you are facing? Perhaps it is the fault of dbi collector, as there are no pings received from dbi in your logs.

**Comment by daledude** [comment]
@mkuculyma, I thought no pings was odd but was thinking dbi was so old that maybe it was a new snapd thing. You can find all my configs [here](https://gist.github.com/daledude/ba46f4e2f6250bdd34d43e3b0e2f8ad8):

**Comment by daledude** [comment]
I think I know the issue but not the cause yet. Telegraf also has this behaviour. The time that snap sees becomes 5 minutes into the future for some reason. Running a loop over the date command doesn't show this issue. So, my guess is the time function snap and telegraf use are possibly a problem with my old kernel. Collectd doesn't seem to have this issue.

What function does snap use for the timestamp?

**Comment by IzabellaRaulin** [comment]
It looks that only logs are affected - we use https://github.com/Sirupsen/logrusto logging and it has own timestamp. If you would like to look more into the code:
- logging "Ping received" happens here:
    -> https://github.com/intelsdi-x/snap/blob/master/control/plugin/session_deprecated.go#L122
- setting lastPing timestamp happens here:
     -> https://github.com/intelsdi-x/snap/blob/master/control/plugin/session_deprecated.go#L164

**Comment by mkuculyma** [comment]
@daledude could you please let us know if you are still experiencing this issue after kernel upgrade? 🐢

**Comment by daledude** [comment]
@mkuculyma upgrading the kernel is a sensitive issue. But, it does appear a bug before RHEL kernel 2.6.18-308 with clock_gettime().

Thanks.
